CI: Cancel redundant workflows

Julow opened the pull request:
The new [concurrency](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency) option allows to group concurrent runs by a key and then to cancel running jobs before running the workflow again.
This avoids starting many CI jobs for which the result is uninteresting.

I'm submitting this because I just force-pushed 4 times on one of my PR (rebase, update changes, mistake in changes, test failure) and I feel like many Wh were wasted.
In my testing, the workflows run after a push are delayed by several seconds (at least 10s, perhaps the time needed to cancel the running jobs) but perhaps this is compensated by the workflows running faster ? (I noticed on an other project that very long jobs get slower if many are run concurrently)

Julow committed:
CI: Cancel redundant workflows

The new 'concurrency' option allow to group concurrent runs by a key and
then to cancel running jobs before running the workflow again.

This avoids starting many CI jobs for which the result is uninteresting.

Julow committed:
Apply to the hygiene workflow as well

gasche commented:
Naive question: when does `cancel` happen? (I wish this was explained in a comment in the code, which could also include the link to the 'concurrency' documentation.)

gasche commented:
(Note: I'm certainly not volunteering to review this, oh no no, this is for @dra27.)

Julow committed:
CI: Add some explanations

Julow commented:
The cancel happens when a new workflow is triggered (eg. a push) but before any job is running because of this trigger, which will be started only after the cancel is finished (it takes time). I added a comment in the code.

abbysmal commented on .github/workflows/build.yml:
```diff
@@ -17,6 +17,16 @@ permissions: {}
+# Concurrent workflows are grouped by the PR or branch that triggered them
+# (github.ref) and the name of the workflow (github.workflow). The
+# 'cancel-in-progress' option then make sure that only one workflow is running
+# at a time. This doesn't prevent new jobs from running, rather is cancels
```
s/is/it/ ?

Julow replied:
Fixed, thanks.

dra27 commented on .github/workflows/build.yml:
```diff
@@ -17,6 +17,16 @@ permissions: {}
+# at a time. This doesn't prevent new jobs from running, rather it cancels
+# already running jobs before scheduling new jobs.
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
```
This tweak allows branch pushes _not_ to cancel each other - on the occasions when a break is discovered after a merge, it's mildly useful to have the first one failing, rather than cancelled.
```suggestion
  group: ${{ github.workflow }}-${{ github.ref }}-${{ github.event_name == 'pull_request' || github.sha }}
```

dra27 reviewed (commented):
Thanks for this! It's a feature we already use on AppVeyor.

It's unquestionably the case that cancelling the running jobs of a PR is a good idea when the PR is changed. For push events (i.e. when a PR has been merged), I think it's actually better, given the choice, _not_ to cancel any existing builds on that branch. GitHub does record that status against the commit on `trunk`, etc. and it is useful if a breakage is only discovered after merging (which does happen) for the _first_ commit which broke the build to be marked as failing.

Julow committed:
Avoid cancelling jobs outside of a pull request

This tweak allows branch pushes not to cancel each other - on the occasions when a break is discovered after a merge, it's mildly useful to have the first one failing, rather than cancelled.

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

Julow commented:
Thanks! I thought `github.ref` was covering this case but I didn't actually tested it.

dra27 reviewed (approved):
LGTM, thanks!
